parser/src/cursor.rs

```diff
@@ -0,0 +1,88 @@
+use crate::moves::Move;
+use lexer::token::Token;
+use lexer::token::TokenType::EndOfFile;
+
+use crate::parser::{ParseError, ParseResult};
+
+/// Parser cursor is used by parsers to navigate in the token stream
+#[derive(Debug, Clone)]
+pub(crate) struct ParserCursor<'a> {
+    /// The manipulated tokens
+    tokens: Vec<Token<'a>>,
+    /// current position in the tokens vector.
+    pos: usize,
+}
+
+impl<'a> ParserCursor<'a> {
+    ///Creates a new cursor at position 0 in the given token vector
+    pub fn new(tokens: Vec<Token<'a>>) -> Self {
+        Self { tokens, pos: 0 }
+    }
+
+    ///advance if next token satisfy the given move
+    /// Returns the token where the move ended if it succeeds, or None instead.
+    /// This method will move the current cursor position on where the move ended.
+    pub fn advance(&mut self, mov: impl Move) -> Option<Token<'a>> {
+        let result = mov.apply(|pos| self.at(pos), self.pos);
+
+        if let Some(new_pos) = result {
+            self.pos = new_pos;
+            return Some(self.at(new_pos - 1));
```

Comment:
The `new_pos - 1` seems to contradict the documentation
> Returns the last token before the end of the move if it succeeds?

Reply:
It returns the last token that the move operation covered (=validated). The move returns the position on which the cursor must continue. Thus, move operation returns (B), which is the new position of the cursor, but the cursor will return (B-1) as it's the last token covered by the movement. 
I admit that the doc is unclear